File: src/shared/types/provider.ts
import type { ImageAttachment } from './attachment'

export type ThinkingMode = 'auto' | 'none' | 'openai' | 'deepseek' | 'qwen' | 'anthropic' | 'gemini' | 'grok' | 'openrouter'

export type ThinkingEffort = 'auto' | 'none' | 'minimal' | 'low' | 'medium' | 'high' | 'xhigh' | 'max' | 'custom'

export interface ThinkingConfig {
  enabled: boolean
  mode: ThinkingMode
  effort?: ThinkingEffort
  budgetTokens?: number
}

export interface ModelConfig {
  id: string
  name: string
  /** 最大上下文长度（tokens），0 表示不限制 */
  maxContextTokens: number
  maxInputTokens?: number
  maxOutputTokens?: number
  reasoningCountsAgainstContext?: boolean
  supportsVision?: boolean
  /** 单独覆盖该模型使用的接口协议 */
  apiFormat?: ApiFormat
  /** 单独覆盖该模型使用的思考模式 */
  thinkingMode?: ThinkingMode
  /** 单独覆盖该模型使用的推理强度 */
  thinkingEffort?: ThinkingEffort
  /** 单独覆盖该模型使用的思考 Token 预算 */
  thinkingBudgetTokens?: number | null
}

export interface ModelContextCapabilities {
  contextWindowTokens: number
  maxInputTokens?: number
  maxOutputTokens?: number
  reasoningCountsAgainstContext?: boolean
}

export interface ProviderTokenUsage {
  inputTokens: number
  outputTokens: number
  reasoningTokens?: number
  totalTokens?: number
}

export type ChatProviderErrorCode =
  | 'CONTEXT_OVERFLOW'
  | 'AUTHENTICATION'
  | 'RATE_LIMIT'
  | 'NOT_FOUND'
  | 'NETWORK'
  | 'UNKNOWN'

export type ApiFormat = 'openai' | 'anthropic' | 'gemini'

/** Provider 持久化配置（含加密后的 API Key） */
export interface ProviderConfig {
  id: string
  name: string
  baseUrl: string
  /** 接口格式协议 */
  apiFormat?: ApiFormat
  /** 加密后的 API Key 引用（safeStorage 或 base64） */
  apiKeyRef: string
  /** 加密方式标识 */
  encryption: 'safeStorage' | 'base64' | 'none'
  /** 该 Provider 下配置的模型列表 */
  models: ModelConfig[]
  /** Thinking/reasoning 输出配置 */
  thinking: ThinkingConfig
  enabled: boolean
  createdAt: string
  updatedAt: string
}

/** 前端展示用的 Provider 信息（不含 API Key 明文） */
export interface ProviderInfo {
  id: string
  name: string
  baseUrl: string
  apiFormat?: ApiFormat
  /** API Key 明文 */
  apiKey: string
  models: ModelConfig[]
  thinking: ThinkingConfig
  enabled: boolean
  createdAt: string
}

/** Provider 新建/编辑表单数据 */
export interface ProviderFormData {
  name: string
  baseUrl: string
  apiFormat?: ApiFormat
  apiKey: string
  models: ModelConfig[]
  thinking: ThinkingConfig
}

/** /v1/models 返回的模型信息 */
export interface ModelInfo {
  id: string
  object: string
  created: number
  owned_by: string
}

/** 连接测试结果 */
export interface ConnectionTestResult {
  success: boolean
  message: string
  models?: string[]
}

/** 聊天消息（与 OpenAI 格式对齐） */
export interface ChatMessage {
  role: 'system' | 'user' | 'assistant' | 'tool'
  content?: string
  tool_calls?: ToolCall[]
  tool_call_id?: string
  name?: string
  attachments?: ImageAttachment[]
}

export interface ToolCall {
  id: string
  type: 'function'
  function: {
    name: string
    arguments: string
  }
  thought_signature?: string
}

export interface ToolDefinition {
  type: 'function'
  function: {
    name: string
    description: string
    parameters: Record<string, any> // JSON Schema
  }
}

export type AgentStopReason = 'stop' | 'length' | 'tool_calls' | 'content_filter' | 'error' | 'unknown'

export interface ToolResult {
  ok: boolean
  data?: any
  error?: string
}

/** 流式 chunk */
export interface ChatStreamChunk {
  /** 本次增量文本 */
  delta: string
}

/** 流式结束 */
export interface ChatStreamEnd {
  /** 完整消息内容 */
  fullContent: string
  /** token 用量（可选） */
  usage?: ProviderTokenUsage
}

File: crates/codez-core/src/provider.rs
use std::{fmt, future::Future, pin::Pin};

use serde::{Deserialize, Serialize};
use thiserror::Error;
use zeroize::Zeroizing;

use crate::PortFuture;

const MAX_CREDENTIAL_KEY_BYTES: usize = 192;
const CREDENTIAL_PREFIX: &str = "provider-api-key";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ThinkingMode {
    Auto,
    None,
    Openai,
    Deepseek,
    Qwen,
    Anthropic,
    Gemini,
    Grok,
    Openrouter,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ThinkingEffort {
    Auto,
    None,
    Minimal,
    Low,
    Medium,
    High,
    Xhigh,
    Max,
    Custom,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ThinkingConfig {
    pub enabled: bool,
    pub mode: ThinkingMode,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub effort: Option<ThinkingEffort>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub budget_tokens: Option<u32>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ApiFormat {
    Openai,
    Anthropic,
    Gemini,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModelConfig {
    pub id: String,
    pub name: String,
    pub max_context_tokens: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_input_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_output_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reasoning_counts_against_context: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub supports_vision: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_format: Option<ApiFormat>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thinking_mode: Option<ThinkingMode>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thinking_effort: Option<ThinkingEffort>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thinking_budget_tokens: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderTokenUsage {
    pub input_tokens: u32,
    pub output_tokens: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reasoning_tokens: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub total_tokens: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ProviderConfig {
    pub id: String,
    pub name: String,
    pub base_url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_format: Option<ApiFormat>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub credential_id: Option<String>,
    pub models: Vec<ModelConfig>,
    pub thinking: ThinkingConfig,
    pub enabled: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderInfo {
    pub id: String,
    pub name: String,
    pub base_url: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub api_format: Option<ApiFormat>,
    pub api_key_configured: bool,
    pub models: Vec<ModelConfig>,
    pub thinking: ThinkingConfig,
    pub enabled: bool,
    pub created_at: String,
}

pub struct ProviderFormData {
    pub name: String,
    pub base_url: String,
    pub api_format: Option<ApiFormat>,
    pub api_key: Option<SecretValue>,
    pub models: Vec<ModelConfig>,
    pub thinking: ThinkingConfig,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ConnectionTestResult {
    pub success: bool,
    pub message: String,
    pub models: Option<Vec<String>>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProvidersFile {
    pub providers: Vec<ProviderConfig>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub active_provider_id: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AgentStopReason {
    Stop,
    Length,
    ToolCalls,
    ContentFilter,
    Error,
    Unknown,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Role {
    System,
    User,
    Assistant,
    Tool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatMessage {
    pub role: Role,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<Vec<ToolCall>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    /// Verified image content used only while building a provider request.
    ///
    /// The desktop runtime rehydrates these bytes from the session attachment
    /// store. They are deliberately excluded from persisted and generic wire
    /// representations.
    #[serde(skip, default)]
    pub images: Vec<ChatImage>,
}

/// Verified image content available to a provider request.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ChatImage {
    pub mime_type: String,
    pub bytes: Vec<u8>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolCall {
    pub id: String,
    pub r#type: String,
    pub function: ToolCallFunction,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub thought_signature: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolCallFunction {
    pub name: String,
    pub arguments: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolDefinition {
    pub r#type: String,
    pub function: ToolDefinitionFunction,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolDefinitionFunction {
    pub name: String,
    pub description: String,
    pub parameters: serde_json::Value,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ChatStreamEvent {
    Chunk {
        delta: String,
        reasoning_delta: Option<String>,
        tool_calls: Option<Vec<ToolCall>>,
        thought_signature: Option<String>,
    },
    Done {
        full_content: String,
        stop_reason: Option<AgentStopReason>,
        tx_id: Option<String>,
    },
    Usage(ProviderTokenUsage),
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct CredentialId {
    provider_id: String,
}

impl CredentialId {
    pub fn new(provider_id: impl Into<String>) -> Result<Self, CredentialError> {
        let provider_id = provider_id.into();
        if provider_id.is_empty()
            || provider_id.len() > MAX_CREDENTIAL_KEY_BYTES
            || !provider_id
                .bytes()
                .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'.' | b'_' | b'-'))
        {
            return Err(CredentialError::InvalidIdentifier);
        }
        Ok(Self { provider_id })
    }

    pub fn parse(value: &str) -> Result<Self, CredentialError> {
        let (prefix, provider_id) = value
            .split_once(':')
            .ok_or(CredentialError::InvalidIdentifier)?;
        if prefix != CREDENTIAL_PREFIX {
            return Err(CredentialError::InvalidIdentifier);
        }
        Self::new(provider_id)
    }

    #[must_use]
    pub fn provider_id(&self) -> &str {
        &self.provider_id
    }

    #[must_use]
    pub fn account_name(&self) -> String {
        format!("{CREDENTIAL_PREFIX}:{}", self.provider_id)
    }
}

impl fmt::Display for CredentialId {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.account_name())
    }
}

pub struct SecretValue(Zeroizing<String>);

impl SecretValue {
    pub fn new(value: impl Into<String>) -> Result<Self, CredentialError> {
        let value = value.into();
        if value.is_empty() {
            return Err(CredentialError::EmptySecret);
        }
        Ok(Self(Zeroizing::new(value)))
    }

    #[must_use]
    pub fn expose_secret(&self) -> &str {
        self.0.as_str()
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Error)]
pub enum CredentialError {
    #[error("credential identifier is invalid")]
    InvalidIdentifier,
    #[error("credential secret must not be empty")]
    EmptySecret,
    #[error("credential was not found: {id}")]
    NotFound { id: CredentialId },
    #[error("credential store denied access while attempting to {operation}")]
    AccessDenied { operation: &'static str },
    #[error("credential store is unavailable while attempting to {operation}")]
    Unavailable { operation: &'static str },
    #[error("stored credential is corrupt: {id}")]
    Corrupt { id: CredentialId },
    #[error("credential secret exceeds the platform length limit of {platform_limit}")]
    SecretTooLarge { platform_limit: u32 },
}

pub type CredentialFuture<'a, T> =
    Pin<Box<dyn Future<Output = Result<T, CredentialError>> + Send + 'a>>;

pub trait CredentialStore: Send + Sync {
    fn get(&self, id: CredentialId) -> CredentialFuture<'_, SecretValue>;
    fn set(&self, id: CredentialId, value: SecretValue) -> CredentialFuture<'_, ()>;
    fn delete(&self, id: CredentialId) -> CredentialFuture<'_, ()>;
}

pub trait ProviderRepository: Send + Sync {
    fn load(&self) -> PortFuture<'_, Option<ProvidersFile>>;
    fn save(&self, data: ProvidersFile) -> PortFuture<'_, ()>;
}
